### ISSUE TITLE
Reordering arrays with proper keys triggers a re-render to the DOM

### ISSUE THREAD

**Issue by ostrgard** [facebook/react#9478]
Hi!

I ran across some weird behavior while implementing some autoplaying video on a recent project. The video is part of an array of modules, and when this array was reordered, the video would stop in Safari on Mac, only to be rendered again. I were eventually able to reproduce this "resetting" behaviour consistently between all browsers with CSS animations and transitions.

This simple example shows the issue quite well: https://jsfiddle.net/ostrgard/rL33d4wL/

Each of the numbers has an animation that runs once, that animates the element 100px to the right, and back again. You'll see this when the page loads. This animation should never be replayed... unless I've completely misunderstood something. Each number is also set as the key.

The buttons reorders the array. Clicking from "1,2,3" to "1,3,2" resets the animation for "2". Clicking from "1,2,3" to "3,2,1" resets the animation for "2" and "1". 

What is happening here? I'd expect the behaviour to never reset an animation or otherwise cause the browser to remove an element and insert a new one.

A few pointers:
1. ~~In Chrome DOM inspector, dragging one element node above the other does not reset animation.~~
2. Changing the items to a React component and logging it's unmount proves there's no unmounting. This seems to be outside of the component lifecycle.
3. I first encountered this issue with React 15.3.0 and the fiddle is using 15.5.4.

Thanks for reading. I'm looking forward to hear your thoughts and learn something from this :)

PS. I am amazed, that every other browser besides Safari is able to continue the video with this behaviour. I suspect it has something to do with Safari using Quicktime.

**Comment by gaearon** [comment]
Have you tried to create a similar example with plain DOM?

**Comment by ostrgard** [comment]
@gaearon I have not, and now you ask I realize that it would be quite a challenge. Can I ask why you ask? I'd like to give it a shot, if you think it could help.

**Comment by gaearon** [comment]
>Can I ask why you ask?

So that we can tell if it’s an issue with React, or a browser bug.

>now you ask I realize that it would be quite a challenge

I don’t think it necessarily would be. Create three DOM nodes, and write the code that reorders them on button click.

**Comment by syranide** [comment]
I'm quite sure this is just another instance of reordering iframes (which resets them). Any time you move an element in the DOM it is basically removed and added, this causes iframes to reset, video to stop, audio stop, etc. Probably also animations to reset. So unless shown otherwise I would assume this is just browsers being intentionally quirky.

**Comment by ostrgard** [comment]
I was wrong about Chrome element devtools drag n' drop not triggering this behaviour. I'm sorry about the confusion. Also, I already implemented a workaround for the video pausing in ie. Chrome, as you are talking about,  @syranide. Had I payed attention to that, I might have understood this issue earlier. So this is indeed the same behaviour. 

@gearon Here's a simple demo: https://jsfiddle.net/ostrgard/92Lhswd5/. You're right, as to this is a browser issue. I was kind of confused about Reacts logic about moving these elements in the DOM, but it's pretty clear to me now that it, maybe too simply put, appends nodes, and it is the browsers that exhibit this behaviour.

Thank you for your help. 

A related question, without too much knowledge about how React actually moves these elements around: Is it possible to do this in a more intelligent way, that limits the number of elements moved to limit this side-effect? I'm thinking of different combinations of append, prepend, insertBefore, insertAfter etc.?
